### PR TITLE
Macro input future proofing

### PR DESCRIPTION
**Pull request by emberian**
Implements RFC 550 (https://github.com/rust-lang/rfcs/pull/550)

### CODE REVIEW

File: src/libsyntax/ext/tt/macro_rules.rs [hunk 1]
```diff
@@ -274,3 +278,131 @@ pub fn compile<'cx>(cx: &'cx mut ExtCtxt,
...
+
+    check_matcher(cx, matcher, &Eof);
+    // we don't abort on errors on rejection, the driver will do that for us
+    // after parsing/expansion. we can report every error in every macro this way.
+}
+
+fn check_matcher(cx: &mut ExtCtxt, matcher: &[TokenTree], follow: &Token) {
+    use print::pprust::token_to_string;
+
+    // 1. If there are no tokens in M, accept
```
Comment (sfackler) [inline src/libsyntax/ext/tt/macro_rules.rs]: This case doesn't need to exist, does it?

File: src/libsyntax/ext/tt/macro_rules.rs [hunk 2]
```diff
@@ -274,3 +278,131 @@ pub fn compile<'cx>(cx: &'cx mut ExtCtxt,
...
+            true
+        },
+        "block" => {
+            // anything can follow block, the braces provide a easy boundary to
+            // maintain
+            true
+        },
+        "stmt" | "expr"  => {
+            match *tok {
+                Comma | Semi => true,
```
Comment (sfackler) [inline src/libsyntax/ext/tt/macro_rules.rs]: Could we have at least one of FatArrow or RArrow here? Are we planning on using them in future expression syntax?

**Comment by nikomatsakis** [comment]
OK, reviewed. Algorithm seems close but not quite right, particularly around adjacent non-terminals (which can't be soundly permitted without a `FIRST` set) and the repeaters.

File: src/libsyntax/ext/tt/macro_rules.rs (hunk 2 above)
Comment (nikomatsakis) [inline src/libsyntax/ext/tt/macro_rules.rs]: No immediate plans. It'd probably be nice to give *some* kind of delimiter beyond `,` or `;`, it's true.

**Comment by nikomatsakis** [comment]
@cmr I'm ok with `=>` in the follow sets of types and expressions. @alexcrichton points out that for expressions we kind of are stuck with it anyway because of `_ if foo =>` in patterns. Also, given that we want to add type ascription, an expression can always end in a type, so it makes sets for the follow sets of the two to be the same.

**Comment by nikomatsakis** [comment]
Sorry, that was too strong. And we don't need type ascription, we already have `expr as Type`. So really just `follow(expr) <= follow(type)` (I think), and since `=>` is in `follow(expr)`, should be present for type as well.

**Comment by SimonSapin** [comment]
I use this macro a lot:
```rust
macro_rules! is_match {
    ($value:expr, $($pattern:pat)|+) => {
        match $value {
            $($pattern)|+ => true,
            _ => false
        }
    }
}
```

But it now gives `error: `$pattern:pat` is followed by `|`, which is not allowed for `pat` fragments`. Why is this specific pair not allowed? Is there a work-around?

**Comment by SimonSapin** [comment]
For what it’s worth, upgrading rust-url to the latest nightly is blocked on this.

**Comment by cmr** [comment]
@SimonSapin that combination is disallowed to provide for extending patterns with `|`-alternation in the future. Right now, `|` in "pattern" is handled entirely by `match`.

The only tokens allowed to follow `pat` at the moment are `=>`, `,`, and `=`.

**Comment by SimonSapin** [comment]
@cmr, so there is no way to write a macro the accepts multiple patterns with the same syntax as `match`? I don’t care if it’s with `$($pattern:pat)|+` or with a single `$pattern:pat`, but making it impossible is a pretty bad regression.

**Comment by alexcrichton** [comment]
@SimonSapin thanks for pointing this out! We were deliberately quite conservative in choosing the "follow sets" for this RFC as we can always expand them over time! For patterns specifically, I think we *must* allow `|` in the follow set because we could never extend the grammar of a pattern to include a `|` token (it would cause many ambiguities today).

If you'd like, you could open a PR modifying [the relevant code](https://github.com/rust-lang/rust/blob/2e2372c6c4a265992b0eb615ea9fdee4ab999143/src/libsyntax/ext/tt/macro_rules.rs#L434-L439) to have a bit more discussion (someone should validate my thinking)

**Comment by SimonSapin** [comment]
Ok, @huonw suggested a work around that I’m happy with:

```rust
#[macro_export]
macro_rules! matches {
    ($expression: expr, $($pattern:tt)+) => {
        _tt_as_expr_hack! {
            match $expression {
                $($pattern)+ => true,
                _ => false
            }
        }
    }
}


/// Work around "error: unexpected token: `an interpolated tt`", whatever that means.
#[macro_export]
macro_rules! _tt_as_expr_hack {
    ($value:expr) => ($value)
}
```

`tt` allows sneaking another `match` arm in, which isn’t great, but better than picking a random ASCII character for the delimiter.

```rust
matches!("foo", "bar" | "baz" => 42 > 9000, "fuzz");
```

**Pull request merged**